File: 2015/Day5/Day5.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <algorithm>

using namespace std;

bool stringContains(string mainString, string subString);

int main(){

	int numNiceStrings = 0;
	ifstream inputFile ("input.txt");

	string stringToCheck;

	while(getline(inputFile, stringToCheck)){

		// Contains at least three vowels
		string const vowels("aeiouAEIOU");
		int vowelCount = count_if(stringToCheck.begin(), stringToCheck.end(), 
			[&](char c){ return vowels.end() != find(vowels.begin(), vowels.end(), c); });

		if (vowelCount < 3){
			continue;
		}

		// Contains at least one letter that appears twice in a row
		bool twiceInRow = false; 
		for (int i = 0; i < stringToCheck.length()-1; i++){
			if(stringToCheck[i] == stringToCheck[i+1]){
				twiceInRow = true;
				break;
			}
		}

		if (!twiceInRow){
			continue;
		}

		// Does not contain strings: ab, cd, pq, or xy
		if (stringContains(stringToCheck, "ab") || stringContains(stringToCheck, "cd") || 
			stringContains(stringToCheck, "pq") || stringContains(stringToCheck, "xy")){
			continue;
		}

		numNiceStrings++;

	}

	cout << "Number of nice strings: " << numNiceStrings << endl;

	return 0;
}


bool stringContains(string mainString, string subString){
	if (mainString.find(subString) != string::npos){
		return true;
	}
	else {
		return false;
	}
}

File: 2015/Day5/Day5.py
import re

def main():

	n_nice_strings = 0;

	with open("input.txt", 'r') as input_file:

		for line in input_file:
			
			# Contains at least three vowels
			if len(re.findall("[aeiouAEIOU]", line)) < 3:
				continue

			# Contains at least one letter that appears twice in a row
			if not re.search("(.)\\1", line):
				continue	

			# Does not contain the strings: ab, cd, pq, or xy
			if re.search("ab|cd|pq|xy", line):
				continue

			n_nice_strings += 1

	print("Number of nice strings:", n_nice_strings)

main()